File: FBControlCore/Configuration/FBInstrumentsConfiguration.m
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBInstrumentsConfiguration.h"

#import "FBCollectionInformation.h"

@implementation FBInstrumentsTimings

#pragma mark Initializers

+ (instancetype)timingsWithTerminateTimeout:(NSTimeInterval)terminateTimeout launchRetryTimeout:(NSTimeInterval)launchRetryTimeout launchErrorTimeout:(NSTimeInterval)launchErrorTimeout operationDuration:(NSTimeInterval)operationDuration
{
  return [[self alloc] initWithterminateTimeout:terminateTimeout launchRetryTimeout:launchRetryTimeout launchErrorTimeout:launchErrorTimeout operationDuration:operationDuration];
}

- (instancetype)initWithterminateTimeout:(NSTimeInterval)terminateTimeout launchRetryTimeout:(NSTimeInterval)launchRetryTimeout launchErrorTimeout:(NSTimeInterval)launchErrorTimeout operationDuration:(NSTimeInterval)operationDuration
{
  self = [self init];
  if (!self) {
    return nil;
  }

  _terminateTimeout = terminateTimeout;
  _launchRetryTimeout = launchRetryTimeout;
  _launchErrorTimeout = launchErrorTimeout;
  _operationDuration = operationDuration;

  return self;
}

@end

@implementation FBInstrumentsConfiguration

#pragma mark Initializers

+ (instancetype)configurationWithTemplateName:(NSString *)templateName targetApplication:(NSString *)targetApplication appEnvironment:(NSDictionary<NSString *, NSString *> *)appEnvironment appArguments:(NSArray<NSString *> *)appArguments toolArguments:(NSArray<NSString *> *)toolArguments timings:(FBInstrumentsTimings *)timings
{
  return [[self alloc] initWithTemplateName:templateName targetApplication:targetApplication appEnvironment:appEnvironment appArguments:appArguments toolArguments:toolArguments timings:timings];
}

- (instancetype)initWithTemplateName:(NSString *)templateName targetApplication:(NSString *)targetApplication appEnvironment:(NSDictionary<NSString *, NSString *> *)appEnvironment appArguments:(NSArray<NSString *> *)appArguments toolArguments:(NSArray<NSString *> *)toolArguments timings:(FBInstrumentsTimings *)timings
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _templateName = templateName;
  _targetApplication = targetApplication;
  _appEnvironment = appEnvironment;
  _appArguments = appArguments;
  _toolArguments = toolArguments;
  _timings = timings;
  return self;
}

#pragma mark NSObject

- (NSString *)description
{
  return [NSString stringWithFormat:
    @"Instruments %@ | %@ | %@ | %@ | %@ | duration %f | terminate timeout %f | launch retry timeout %f | launch error timeout %f",
    self.templateName,
    self.targetApplication,
    [FBCollectionInformation oneLineDescriptionFromDictionary:self.appEnvironment],
    [FBCollectionInformation oneLineDescriptionFromArray:self.appArguments],
    [FBCollectionInformation oneLineDescriptionFromArray:self.toolArguments],
    self.timings.operationDuration,
    self.timings.terminateTimeout,
    self.timings.launchRetryTimeout,
    self.timings.launchErrorTimeout
  ];
}

#pragma mark NSCopying

- (instancetype)copyWithZone:(NSZone *)zone
{
  return self;
}

@end

File: FBControlCore/Configuration/FBInstrumentsConfiguration.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation

public struct FBInstrumentsTimings: Sendable {

  public let terminateTimeout: TimeInterval
  public let launchRetryTimeout: TimeInterval
  public let launchErrorTimeout: TimeInterval
  public let operationDuration: TimeInterval

  public static func timings(withTerminateTimeout terminateTimeout: TimeInterval, launchRetryTimeout: TimeInterval, launchErrorTimeout: TimeInterval, operationDuration: TimeInterval) -> FBInstrumentsTimings {
    FBInstrumentsTimings(terminateTimeout: terminateTimeout, launchRetryTimeout: launchRetryTimeout, launchErrorTimeout: launchErrorTimeout, operationDuration: operationDuration)
  }

  public init(terminateTimeout: TimeInterval, launchRetryTimeout: TimeInterval, launchErrorTimeout: TimeInterval, operationDuration: TimeInterval) {
    self.terminateTimeout = terminateTimeout
    self.launchRetryTimeout = launchRetryTimeout
    self.launchErrorTimeout = launchErrorTimeout
    self.operationDuration = operationDuration
  }
}

public struct FBInstrumentsConfiguration: Sendable, CustomStringConvertible {

  public let templateName: String
  public let targetApplication: String
  public let appEnvironment: [String: String]
  public let appArguments: [String]
  public let toolArguments: [String]
  public let timings: FBInstrumentsTimings

  public static func configuration(withTemplateName templateName: String, targetApplication: String, appEnvironment: [String: String], appArguments: [String], toolArguments: [String], timings: FBInstrumentsTimings) -> FBInstrumentsConfiguration {
    FBInstrumentsConfiguration(templateName: templateName, targetApplication: targetApplication, appEnvironment: appEnvironment, appArguments: appArguments, toolArguments: toolArguments, timings: timings)
  }

  public init(templateName: String, targetApplication: String, appEnvironment: [String: String], appArguments: [String], toolArguments: [String], timings: FBInstrumentsTimings) {
    self.templateName = templateName
    self.targetApplication = targetApplication
    self.appEnvironment = appEnvironment
    self.appArguments = appArguments
    self.toolArguments = toolArguments
    self.timings = timings
  }

  public var description: String {
    "Instruments \(templateName) | \(targetApplication) | \(FBCollectionInformation.oneLineDescription(from: appEnvironment)) | \(FBCollectionInformation.oneLineDescription(from: appArguments)) | \(FBCollectionInformation.oneLineDescription(from: toolArguments)) | duration \(timings.operationDuration) | terminate timeout \(timings.terminateTimeout) | launch retry timeout \(timings.launchRetryTimeout) | launch error timeout \(timings.launchErrorTimeout)"
  }
}
